[RFC] Normalize `make binary-dist` filenames

staticfloat opened the pull request:
With this change, the output of `make binary-dist` filenames are nicely normalized across platforms.  Running this branch through the buildbots results in the following set of files:

```
julia-19d61d20ff-linux32.tar.gz
julia-19d61d20ff-linux64.tar.gz
julia-19d61d20ff-linuxarm.tar.gz
julia-19d61d20ff-linuxppc64le.tar.gz
julia-19d61d20ff-mac64.dmg
julia-19d61d20ff-win64.exe *
julia-19d61d20ff-win32.exe *
```

We should get the two windows ones (that have a `*` after them) as well, but I'm running into weird windows errors while trying to build llvm, so I can't _actually_ verify that works.

staticfloat committed:
Make Julia binary dist filenames consistent across platforms

tkelman commented:
we may eventually have aarch64 buildbots?

windows llvm error should be fixed if you rebase against latest master

tkelman commented:
merging this would need updates to nightly links in a number of places around the ecosystem, so please wait until we've identified where those all are.

tkelman commented:
since we're not going to backport this, how can we most easily make the buildbots capable of handling nightlies with new names as well as releases from 0.4 and 0.5 with the old names?

staticfloat commented:
> we may eventually have aarch64 buildbots?

`$(DIST_ARCH)` defaults to `$(ARCH)`, so until we have a solid idea of what we want to suffix the filename with, (is `aarch64` fine?  do we call it `arm64`?) I think leaving it the way it is is the best option.

> how can we most easily make the buildbots capable of handling nightlies with new names as well as releases from 0.4 and 0.5 with the old names?

I think the easiest way to do this is to make the buildbots themselves completely agnostic to the build artifact filename.  They will simply run `make print-JULIA_BINARYDIST_FILENAME` store that in e.g. `filename`, then upload `$filename*` to AWS.  This will allow us to upload whatever we want, and if we need to rename 0.4/0.5 pathnames, so be it.  We need to do the bucket dance in any case, so that's not a huge burden for us.

To move this forward, I think we will need to change the stable download buckets to conform to this  naming strategy, (leaving the old links alive for a week or so to let old scripts get updated) and update Travis/Appveyor to use the new download links.

tkelman commented:
I don't think it's okay to remove the old links, ever. We can make copies, will have to see whether that will impact our S3 usage negatively.

staticfloat commented on Make.inc:
```diff
@@ -652,6 +652,32 @@ ifneq (,$(filter $(ARCH), powerpc64 ppc64))
 
+# File name of make binary-dist result
+ifeq ($(JULIA_BINARYDIST_FILENAME),)
+DIST_OS:=$(OS)
```
I like that idea.  I think we can assume that `tr` is available on all these misc. systems.

tkelman replied:
I think someone has defined gmake functions for it too

staticfloat committed:
Lowercase catch-all definition for `DIST_OS`.

tkelman commented:
While we're thinking about renaming a bunch of things, should we also take the opportunity to think about switching from gzip compression to something slightly better like bzip2? Is that suitably ubiquitous by now that it wouldn't be a problem, and save some bandwidth and s3 space?

staticfloat commented:
If we were to switch over to bz2, we'd save about 3% on average.  If we were to switch over to xz, we'd save about 10% on average.

[Here's the script](https://gist.github.com/staticfloat/7f01cf53747001429b8840886ffc4af2) to do the analysis, here's results on our 0.5.0 release, showing unpacked size first, then filename, then gz proportional size, then `bz2` and `xz` proportional sizes, with deltas against `gz`:

```
[110.2MB]       0.5.0/julia-0.5.0-full.tar.gz:          gz (80.76%),    bz2 (80.13% - 0.63%),   xz (78.04% - 2.72%)
[158.8MB]       0.5.0/julia-0.5.0-linux-arm.tar.gz:     gz (30.46%),    bz2 (27.51% - 2.95%),   xz (20.04% - 10.42%)
[192.6MB]       0.5.0/julia-0.5.0-linux-i686.tar.gz:    gz (28.62%),    bz2 (25.16% - 3.46%),   xz (18.52% - 10.10%)
[188.3MB]       0.5.0/julia-0.5.0-linux-ppc64le.tar.gz: gz (27.55%),    bz2 (23.81% - 3.74%),   xz (17.16% - 10.39%)
[217.3MB]       0.5.0/julia-0.5.0-linux-x86_64.tar.gz:  gz (27.49%),    bz2 (24.35% - 3.14%),   xz (17.34% - 10.15%)
[26.4MB]        0.5.0/julia-0.5.0.tar.gz:               gz (22.83%),    bz2 (17.87% - 4.96%),   xz (13.78% - 9.05%)
```

tkelman commented:
Yay, data. Probably not worth it then.

staticfloat commented:
I've rebased this, and I'm ready to try pushing this forward once again.

> Though I'd compare relative to the gzip packed sizes as a baseline since that's what we store and what people download. So xz is roughly a third smaller than gz.

You're right, that was a flubbed comparison, and my final conclusion was off by more than a few MB.  I've updated [the analysis script](https://gist.github.com/staticfloat/1a7057ae5b05475d0781f09130feefac) and re-run it to generate the following:

```
[110.2MB]       0.5.0/julia-0.5.0-full.tar.gz:  gz (80.76%),    bz2 (80.13% - 0.63%; 99.23%),     xz (78.04% - 2.72%; 96.64%)
[158.8MB]       0.5.0/julia-0.5.0-linux-arm.tar.gz:     gz (30.46%),    bz2 (27.51% - 2.95%; 90.29%),     xz (20.04% - 10.42%; 65.77%)
[192.6MB]       0.5.0/julia-0.5.0-linux-i686.tar.gz:    gz (28.62%),    bz2 (25.16% - 3.46%; 87.94%),     xz (18.52% - 10.10%; 64.71%)
[188.3MB]       0.5.0/julia-0.5.0-linux-ppc64le.tar.gz: gz (27.55%),    bz2 (23.81% - 3.74%; 86.41%),     xz (17.16% - 10.39%; 62.30%)
[217.3MB]       0.5.0/julia-0.5.0-linux-x86_64.tar.gz:  gz (27.49%),    bz2 (24.35% - 3.14%; 88.56%),     xz (17.34% - 10.15%; 63.08%)
[26.4MB]        0.5.0/julia-0.5.0.tar.gz:       gz (22.83%),    bz2 (17.87% - 4.96%; 78.26%),     xz (13.78% - 9.05%; 60.37%)
```

The third numbers in the parenthesis, (the ones following the semicolon) are measured relative to `gz` compression, so we can see that we save somewhere around 12% using bz2, and somewhere around 37% using `xz`.  Average binary release sizes (e.g. looking at `arm`, `i686`, `ppc64le` and `x86_64`) are:

```
gz: 53.7 MB
bz2: 47.4 MB
xz: 34.3 MB
```

So you can see, we'd actually be saving somewhere around 20 MB by switching over to `xz`, which is quite the difference to what I said above.  However, I don't think we've yet reached a point where we can force anyone that wants to run Julia to use `.xz`, so this is all still moot.  :(

staticfloat commented:
To test this out a little more thoroughly, I'll be running them through the testing buildbots which have been modified to expect these changes in the codebase.

* [OSX build](https://buildtest.e.ip.saba.us/#/builders/1/builds/61)
* [linux64 build](https://buildtest.e.ip.saba.us/#/builders/49/builds/44)
* [linux32 build](https://buildtest.e.ip.saba.us/#/builders/36/builds/45)
* [aarch64 build](https://buildtest.e.ip.saba.us/#/builders/69/builds/19)

staticfloat commented:
Note, OSX build failed because Travis is now routing our builds to the Xcode 7.3 image.

staticfloat committed:
Use `$(JULIA_BINARYDIST_FILENAME)` in `contrib/mac/app/Makefile`

If we define this makevar, we should at least use it where appropriate

staticfloat committed:
Add top-level `make app` makefile target

staticfloat commented:
I've also added a makefile target `app` to the top-level Makefile.  This makes things every so slightly more inline across platforms, our magic incantation for building binary releases is:

```
make win-extras (windows only)
make binary-dist
make app (osx only)
```

After those three steps, in all three cases, there is a file sitting in the top-level directory named `$(JULIA_BINARYDIST_FILENAME).{tar.gz,dmg,exe}`. And no, the `make binary-dist` on OSX is not a complete waste, as `make app` will run `make binary-dist` anyway, and although we redo some work, it's not very much, just tar'ing and untar'ing again.

tkelman commented:
this will need to change contrb/prepare_release.sh as well, I believe

kshyatt commented:
I'm here, being an annoying 🐝  and buzzing you to review this.

tkelman commented:
Not ready until we change the buildbots, which is waiting until after we release 0.5.1.

staticfloat commented:
Now that 0.5.1 is out, this is something I'd like to try and get in so we can make sure it's working on the 0.6-rc's.

tkelman commented:
Doesn't this need the buildbots to be reconfigured in a way that breaks building from release-0.5? I think this has missed the boat for 0.6.

Also still would need to change contrib/prepare_release.sh.

tkelman commented on Makefile:
```diff
@@ -473,13 +468,17 @@ endif
 
+app:
+	$(MAKE) -C contrib/mac/app
+	-mv contrib/mac/app/$(JULIA_BINARYDIST_FILENAME).dmg $(BUILDROOT)
```
don't think we should allow this to fail

staticfloat replied:
That is probably true.  I think I meant `@` instead of `-`.

staticfloat committed:
Makefile: Silence `mv`, don't ignore return value

staticfloat committed:
Update contrib/prepare_release.sh for consistent distnames

tkelman commented:
Can we start with just the new targets needed to make the new buildbot configuration capable of building the old branches? On the buildbot side I'd prefer defaulting to the old names unless such a target exists, otherwise we lose the ability to rebuild any existing release tags.

staticfloat commented:
> Doesn't this need the buildbots to be reconfigured in a way that breaks building from release-0.5? I think this has missed the boat for 0.6.

No, the new buildbot infrastructure can build pre this commit or post this commit just fine.  See [this file](https://github.com/staticfloat/julia-buildbot/blob/level_up/master/package.py#L51-L70) for the details.

> Also still would need to change contrib/prepare_release.sh.

I've taken an initial stab at it, but I haven't tested it by actually running it.

> This would likely also break Travis builds until that gets updated, and that's not something we have direct control over.

As I don't think we should remove old binary names, what this will break is OSX nightlies (linux names should stay the same).  We can maybe fix that quickly enough that it won't be too much of an issue.

tkelman commented:
And what about the existing buildbot infrastructure? Can it handle these renamings without breaking existing branches? Viral is already bothering me about making an 0.5.2, and we're overdue for a 0.6.0 beta tag. Are the new buildbots are ready yet on Mac and Windows? I didn't get the impression we'd be switching prior to 0.6 being done.

staticfloat commented:
> And what about the existing buildbot infrastructure? Can it handle these renamings without breaking existing branches?

Yes, we need to switch buildbots before this will work, I'm not planning on backporting my changes to the old buildbots.

> Are the new buildbots are ready yet on Mac and Windows?

We just need to replace the master, the slaves can be in-place upgrades.  I've tested that with Mac, but not with Windows.  I want to redo how I'm building the Windows buildbots, as we've talked about recently, to get them more in-line with the Linux buildbots, but that isn't holding this up.

> I didn't get the impression we'd be switching prior to 0.6 being done.

If that's the case, then yes, this has missed the boat for 0.6.  Oh well, too bad so sad. :P

staticfloat commented:
To remind ourselves, when this goes in, we'll need to change [these lines](https://github.com/travis-ci/travis-build/blob/f9c59bf4af125e91314586b0a4fca28cf3d7173c/lib/travis/build/script/julia.rb#L90-L115) in Travis.

staticfloat commented:
As our buildbots have stopped working due to the old Cent OS 5.11 buildbots no longer being able to talk to the internet at large due to not having TLS 1.1+ support, I'm effecting my long-overdue buildbot upgrade.  I'm going to merge this over the next day or two, which means nightly names will change slightly.  I'll open a PR against Travis to use the new nightly names once that's working.
